iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp

```diff
@@ -294,6 +308,93 @@ struct TileBatchMatmulPattern
   SmallVector<int64_t, 3> workgroupSize;
 };
 
+/// Computes the Value for subgroupID along each dimension given number of
+/// subgroups `numSubGroups` along each dimension (x-first, y-second, z-third).
+static SmallVector<linalg::ProcInfo, 2> getSubgroupIdsAndCounts(
+    OpBuilder &builder, Location loc, ArrayRef<int64_t> numSubgroups) {
+  Type indexType = builder.getIndexType();
+  Value subgroupId = builder.create<gpu::SubgroupIdOp>(loc, indexType);
+
+  size_t numDims = numSubgroups.size();
+  SmallVector<linalg::ProcInfo, 2> procInfo(numDims);
+  SmallVector<int64_t, 2> procStride(numDims);
+  int64_t linearizedNumProcs = 1;
+  for (size_t i = 0; i < numDims; ++i) {
+    procStride[i] = linearizedNumProcs;
+    linearizedNumProcs *= numSubgroups[i];
+  }
+
+  // subgroupId = id.z * stride_nprocs.z + id.y * stride_nprocs.y + id.x *
+  // stride_nprocs.x
+  Value procId = subgroupId;
+  using namespace edsc::op;
+  edsc::ScopedContext(builder, loc);
+  for (unsigned i = 0; i != numDims; ++i) {
+    Value stride =
+        builder.create<ConstantIndexOp>(loc, procStride[numDims - i - 1]);
+    procInfo[i].procId =
+        builder.create<SignedDivIOp>(loc, indexType, procId, stride);
+    procId = procId % stride;
+    procInfo[i].nprocs =
+        builder.create<ConstantIndexOp>(loc, numSubgroups[numDims - 1 - i]);
+  }
+  return procInfo;
+}
+
+/// The size of the co-operative matrix multiply operations on the device.
+// TODO(ravishankarm, thomasraoux): This needs to be querried from the device.
+Optional<std::array<int64_t, 3>> getCooperativeMatmulSubgroupSize(
```

Comment:
Maybe should be called getMinCooperativeSize()? Since the size of the cooperative matrix can be bigger but we use that as the smallest granularity?